File: BikeSharingdemand.R
rm(list=ls())
#Accept training and test data from command line
args = commandArgs(trailingOnly=TRUE)

# test if there are two argument: if not, return an error
if (length(args)< 1) 
{
  stop("train data must be supplied (ex:day.csv)", call.=FALSE)
}else if (length(args) == 1) 
{
  train_file = args[1]
}


load_libraries = c("ggplot2","corrgram","caret","rpart","MASS","sampling","DMwR","randomForest")

check_installed = function(x){
  if(!(x %in% row.names(installed.packages()))){
    install.packages(x)
  }
}
#install packages if not installed
lapply(load_libraries,check_installed)


lapply(load_libraries,require,character.only = TRUE)

rm(load_libraries)
rm(check_installed)

wd = getwd()
if (!is.null(wd)) setwd(wd)

data = read.csv(train_file)

head(data)

#Remove dteday,instant columns
#Because instant and dteday is not carrying any info
data = subset(data,select = -c(dteday,instant))

##Explore the data
str(data)

#Convert categorical data from numeric to factor
columns_factors = c('season','yr','mnth','holiday','weekday','workingday','weathersit')

data[,columns_factors] = lapply(data[,columns_factors],factor)

#############Missing value analysis######
sum(is.na(data))#There are no missing values. So no need to impute

#############Data visualization##########
pl1 = ggplot(data,aes(x=cnt)) 
pl1 + geom_histogram(fill="blue" )

pl2 = ggplot(data,aes(x=season,y=cnt,fill=yr))
pl2 + geom_bar(stat="summary",position = "dodge",fun.y = "mean")

pl3 = ggplot(data,aes(x=mnth,y=cnt,fill=yr))
pl3 + geom_bar(stat="summary",position = "dodge",fun.y = "mean")

pl4 = ggplot(data,aes(x=weathersit,y=cnt,fill=yr))
pl4 + geom_bar(stat="summary",position = "dodge",fun.y = "mean")

pl5 = ggplot(data,aes(x=weekday,y=cnt,fill=yr))
pl5 + geom_bar(stat="summary",position = "dodge",fun.y = "mean")

pl4 = ggplot(data)
pl4+geom_smooth(aes(x=temp,y=cnt),model = lm)
pl4+geom_smooth(aes(x=hum,y=cnt),model = lm)
pl4+geom_smooth(aes(x=windspeed,y=cnt),model = lm)



####Outlier Analysis#####
# ## BoxPlots - Distribution and Outlier Check
numeric_index = sapply(data,is.numeric) #selecting only numeric

numeric_data = data[,numeric_index]

cnames = colnames(numeric_data)

for (i in 1:length(cnames))
{
 assign(paste0("gn",i), ggplot(aes_string(y = (cnames[i]), x = "weekday"), data = subset(data))+ 
          stat_boxplot(geom = "errorbar", width = 0.5) +
          geom_boxplot(outlier.colour="red", fill = "grey" ,outlier.shape=18,
                       outlier.size=1, notch=FALSE) +
          theme(legend.position="bottom")+
          labs(y=cnames[i],x="weekday")+
          ggtitle(paste("Box plot of weekday for",cnames[i])))
}

## Plotting plots together
gridExtra::grid.arrange(gn1,gn2,gn3,ncol=3)
gridExtra::grid.arrange(gn4,gn5,gn6,gn7,ncol=4)

#df = data
data.frame()
# #loop to remove from all variables
for(i in cnames){
 #print(i)
 val = data[,i][data[,i] %in% boxplot.stats(data[,i])$out]
 print(length(val))
 data[,i][data[,i] %in% val] = NA
}

data = knnImputation(data, k = 3)


#As registered and casual data will not be present 
#as input for prediction. These are considered dependant variables
#and removed for model building
data = subset(data,select = -c(registered,casual))

#####Feature Selection########
## Correlation Plot 
corrgram(data, order = F,
         upper.panel=panel.pie, text.panel=panel.txt, main = "Correlation Plot")


## Chi-squared Test of Independence
factor_index = sapply(data,is.factor)
factor_data = data[,factor_index]

chitestres = matrix(nrow = 7,ncol = 7)

for (i in 1:length(factor_data))
{
  for (j in 1:length(factor_data))
  {
    chitest = chisq.test(table(factor_data[,i],factor_data[,j]))
    if(chitest$p.value < 0.05 && i!=j)
    {
      chitestres[i,j] = chitest$p.value
    }
  }
}

chitest = data.frame(chitestres,row.names = colnames(factor_data))
colnames(chitest) = colnames(factor_data)

#Remove columns
data = subset(data,select = -c(atemp,workingday))


####sampling#####
tr.idx = createDataPartition(data$cnt,p=0.8,list=FALSE)
train = data[tr.idx,]
test = data[-tr.idx,]

#MAPE
#calculate MAPE
MAPE = function(y, pred){
  mean(abs((y - pred)/y))
}

#calculate RMSE
RMSE = function(y,pred){
  sqrt(mean((y-pred)^2))
}

#Linear Regression
#check multicollearity
library(usdm)
vif(data[,7:9])

vifcor(data[,7:9], th = 0.9)

#run regression model
lm_model = lm(cnt ~., data = train)

#Summary of the model
summary(lm_model)

#Predict
predictions_LR = predict(lm_model, test[,1:9])

qplot(x = test[,10], y = predictions_LR, data = test, color = I("blue"), geom = "point")

#Calculate MAPE
MAPE(test[,10], predictions_LR)

#Calculate RMSE
RMSE(test[,10], predictions_LR)

#Calculate other error metrics
regr.eval(test[,10],predictions_LR,stats = c("mae","rmse","mape","mse"))


################Random Forest###############
RF_model = randomForest(cnt ~ ., train, importance = TRUE, ntree = 500)

summary(RF_model)

#Predict test data using random forest model
RF_Predictions = predict(RF_model, test[,-10])

qplot(x = test[,10], y = RF_Predictions, data = test,colour = I("blue"), geom = "point")

#Calculate MAPE
MAPE(test[,10], RF_Predictions)

#Calculate RMSE
RMSE(test[,10],RF_Predictions)

#Calculate other error metrics
regr.eval(test[,10],RF_Predictions,stats = c("mae","rmse","mape","mse"))


#Sample outputs
sample_output = data.frame(test[,10],predictions_LR,RF_Predictions)
colnames(sample_output) = c("true_values","Linear regression pred","Random forest pred")
write.csv(sample_output,"sample_output.csv",row.names = FALSE)

File: BikeSharingdemand.py
import os
import sys
if(len(sys.argv) != 2):
    print("Usage : python BikeSharingdemand.py path\\to\\day.csv ")
    sys.exit(1)
train_file = sys.argv[1]
os.chdir(os.curdir)
os.getcwd()


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
from fancyimpute import KNN  
from scipy.stats import chi2_contingency

data = pd.read_csv("day.csv")

data.head()

#check for null
data.isnull().sum()

data.info()

data.dtypes

data.describe()

data.drop(['instant','dteday'],axis=1,inplace=True)

data.head()

#sns.barplot(x='season',y='cnt',hue='mnth',data=data,estimator=np.std)
sns.barplot(x='season',y='cnt',hue ='yr',data=data)

sns.barplot(x='mnth',y='cnt',data=data,hue='yr')

sns.barplot(x='weathersit',y='cnt',data=data,hue='yr')

sns.barplot(x='weekday',y='cnt',data=data,hue='yr')

#sns.countplot(x='',data=data)

sns.lmplot(x='temp',y='cnt',data=data,col = 'season',row = 'yr')

sns.lmplot(x='hum',y='cnt',data=data,col = 'season',row = 'yr')

sns.lmplot(x='windspeed',y='cnt',data=data,col = 'season',row = 'yr')

#casual and registered are dependent variables, so remove them from model development--> casual + registered => cnt
data.drop(['casual','registered'],axis=1,inplace=True)


# # Outlier analysis
data.head()

#df = data.copy()

#data = df.copy()

cnames = ["temp","atemp","hum","windspeed","cnt"]

fig,ax = plt.subplots(5,figsize = (12,20))
for i in range(0,len(cnames)):
    sns.boxplot(y=cnames[i],x="weekday",data=data,ax=ax[i])

#Detect and delete outliers from data
for i in cnames:
    q75, q25 = np.percentile(data[i], [75 ,25])
    iqr = q75 - q25

    min = q25 - (iqr*1.5)
    max = q75 + (iqr*1.5)

    #Replace with NA
    data.loc[data[i] < min,i] = np.nan
    data.loc[data[i] > max,i] = np.nan

#Calculate missing value
missing_val = pd.DataFrame(data.isnull().sum())

#Impute with KNN
data = pd.DataFrame(KNN(k = 3).fit_transform(data), columns = data.columns)


# # Feature selection
data_corr = data.loc[:,cnames]

#Set the width and height of the plot
f, ax = plt.subplots(figsize=(12, 7))

#Generate correlation matrix
corr = data_corr.corr()

#Plot using seaborn library
#sns.heatmap(corr, mask=np.zeros_like(corr, dtype=np.bool), cmap=sns.diverging_palette(220, 10, as_cmap=True),
#            square=True, ax=ax)
sns.heatmap(corr, mask=np.zeros_like(corr, dtype=np.bool), cmap='coolwarm',square = True,linewidths = 1,ax=ax,annot =True)

cat_names = ["season","yr","mnth","holiday","weekday","workingday","weathersit"]
#loop for chi square values
for i in cat_names:
    for j in cat_names:
        if(i != j):
            chi2, p, dof, ex = chi2_contingency(pd.crosstab(data[i], data[j]))
            if(p < 0.05):
                print(i,"and",j,"are dependants")

#Drop columns
data.drop(['atemp','workingday'],axis=1,inplace=True)

data.head()

data.shape


# # Sampling
# 
#Import Libraries for decision tree
from sklearn import tree
from sklearn.metrics import accuracy_score
from sklearn.cross_validation import train_test_split
from sklearn import metrics

data.head()

data.columns

#Divide data into train and test
X = data.values[:, 0:9]
Y = data.values[:,9]
#X= data[['season', 'yr', 'mnth', 'holiday', 'weekday','weathersit', 'temp', 'hum', 'windspeed', 'casual', 'registered']]
#Y = data['cnt']

x_train, x_test, y_train, y_test = train_test_split( X, Y, test_size = 0.2)


# # Model

# ## Linear regression
from sklearn.linear_model import LinearRegression
model = LinearRegression()

model.fit(x_train,y_train)

model.coef_

pred_LR = model.predict(x_test)

plt.scatter(y_test,pred_LR)

sns.distplot((y_test-pred_LR))

def mape(y_test,pred):
    return np.mean(np.abs((y_test-pred)/y_test))

rms = np.sqrt(np.mean(np.power((np.array(y_test) - np.array(pred_LR)),2)))
rms

metrics.mean_absolute_error(y_test,pred_LR) #mas

metrics.mean_squared_error(y_test,pred_LR) #mse

np.sqrt(metrics.mean_squared_error(y_test,pred_LR)) #rms

metrics.explained_variance_score(y_test,pred_LR)

#mape = np.mean(np.abs((np.array(y_test) - np.array(pred_LR))/np.array(y_test)))
#mape
mape(y_test,pred_LR)


# ## Random forest
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor()

model.fit(x_train,y_train)

pred_RF = model.predict(x_test)

plt.scatter(y_test,pred_RF)

sns.distplot((y_test-pred_RF))

rms = np.sqrt(np.mean(np.power((np.array(y_test) - np.array(pred_RF)),2)))
rms

metrics.mean_absolute_error(y_test,pred_RF) #mas

np.sqrt(metrics.mean_squared_error(y_test,pred_RF)) #rms

metrics.explained_variance_score(y_test,pred_RF)

#mape = np.mean(np.abs((np.array(y_test) - np.array(pred))/np.array(y_test)))
#mape
mape(y_test,pred_RF)

np.mean(np.abs((np.array(y_test) - np.array(pred_RF))/np.array(y_test)))


# # Sample input and output to excel file
#pd.DataFrame([y_test,pred_LR,pred_RF])
#pd.DataFrame
data = {'True values': y_test, 'Linear regression pred': pred_LR,'Random Forest pred': pred_RF}
sampleoutput = pd.DataFrame(data)

sampleoutput.to_csv("predicted_values.csv")

#pd.DataFrame(x_test,columns = ['season', 'yr', 'mnth', 'holiday', 'weekday', 'weathersit', 'temp',
#       'hum', 'windspeed']).to_csv("test_data.csv")

